packages/vuelidate/index.d.ts

```diff
@@ -40,7 +40,7 @@ export interface ValidatorResponse {
   [key: string]: any
 }
 
-export type ValidatorFn <T = unknown> = (value: T) => boolean | ValidatorResponse | Promise<boolean | ValidatorResponse>;
+export type ValidatorFn <T = unknown> = (value: T, vm: Vue) => boolean | ValidatorResponse | Promise<boolean | ValidatorResponse>;
```

Comment:
Is this always correct? I'm checking this in a Vue 2.6 app using the Composition API library and the type of `vm` appears to be `Vue & ThingsAvailableToTheTemplate`, where `ThingsAvailableToTheTemplate` is dependent on the component itself. `Vue` by itself doesn't have the appropriate scope.

Reply:
the second parameter is the component's Vue instance yeah.

Reply:
Yes, but the type wouldn’t be `Vue` since that gives you something more general, the TypeScript compiler doesn’t know that it represents this actual component. I haven’t come across a type in Vue that does this for you, I think you need a second generic argument, something like `V extends Vue = Vue`, that will allow the user to describe it. Even that might be too restrictive, so something like `V = any` will at least allow the user to handle it while a more appropriate type is figured out.

Reply:
ahaa, well I am still a TS noob :/ Would you care making a PR so we can clean such things up?

Reply:
what about specifying the second param as a generic which extends the `Vue` component
`export type ValidatorFn <T = unknown, K extends Vue> = (value: T, vm: K) => boolean | ValidatorResponse | Promise<boolean | ValidatorResponse>;
`

Reply:
@basti-n I think that's the winning move since it would let the user just feed the function the things they know that the compiler couldn't know otherwise.

I think the use of `unknown` here is a problem, too. This fails to compile:

```typescript
  setup(props) {
    const specialRule: ValidatorFn<string | null> = (input) => {
      const confidentAboutInput = input as string | null;
      return !!confidentAboutInput && confidentAboutInput !== 'foo';
    };

    const rules = {
      target: {
        value: {
          specialRule
        }
      }
    };

    const vvv = useVuelidate(rules, props);
  }
```

![B6C57E7E-5D12-4E4C-A0EB-C7F8F8BE41D2](https://user-images.githubusercontent.com/4097271/133646082-e1c2f0c7-8d7d-4802-97d2-6ea7afe4a7f9.png)

The definition of my special rule is fine but `useValidate` is failing. This happens because `useVuelidate` has these types for its first parameter:

```typescript
validationsArgs: Ref<Vargs> | ValidationArgs,

# which refers to...

export interface ValidationArgs {
  [K: string]: ValidationRule | ValidationArgs
}

# which refers to...

export type ValidationRule <T = unknown> = ValidationRuleWithParams<any, T> | ValidationRuleWithoutParams<T> | ValidatorFn<T>;

# which refers to...

export type ValidatorFn <T = unknown> = (value: T) => boolean | ValidatorResponse | Promise<boolean | ValidatorResponse>;
```

Since `ValidationArgs` is passed without a generic no matter what the user does when calling `useVuelidate`, it means that it's going to compare my deliberately typed `specialRule: ValidatorFn<string | null>` to `ValidatorFn<unknown>`. I'd probably swap those `T = unknown` for `T = any` (or would `T extends unknown` do the trick?) to at least let users take control over it and avoid this compiler error. Alternatively, you could write a type that extracts the value of T from any `ValidatorFn`s passed into `validationsArgs`, but I'm not sure what that would take.

@dobromir-hristov, I wasn't able to get a good dev environment working for this library with TypeScript. If anyone's got an hour free sometime soon and wants to do some remote pairing on it, I'm happy to jump in.

Reply:
Having the value of T extracted certainly is better from a type-correctness PoV. For the time being, and given that typing with `any` seems not to be prohibited by linting going with `any` c/would also be fine. 

For your case, why don't you keep the `ValidatorFn<unknown>` and have the type narrowed to string or null within the specialRule? 

```
const isStringOrNull = (val: unknown): val is number | null => {
    return typeof val === 'string' || val === null
}
```

Reply:
We're doing something like that as a workaround but it seems like the intention is for `ValidatorFn<string | unknown>` to let us avoid doing that. I think use of `any` + your `Record` example above would be a good solution to unblock folks for now while more permanent types are explored.